homeassistant/components/vacuum/__init__.py

```diff
@@ -0,0 +1,361 @@
+            charging = 'charg' in self.status.lower()
+        return icon_for_battery_level(
+            battery_level=self.battery_level, charging=charging)
+
+    @property
+    def fan_speed(self):
+        """Return the fan speed of the vacuum cleaner."""
+        return None
+
+    @property
+    def fan_speed_list(self) -> list:
+        """Get the list of available fan speed steps of the vacuum cleaner."""
+        return []
+
+    @property
+    def state_attributes(self):
+        """Return the state attributes of the vacuum cleaner."""
+        data = {}
+
+        if self.status is not None:
+            data[ATTR_STATUS] = self.status
+
+        if self.battery_level is not None:
+            data[ATTR_BATTERY_LEVEL] = self.battery_level
+            data[ATTR_BATTERY_ICON] = self.battery_icon
+
+        if self.fan_speed is not None:
+            data[ATTR_FAN_SPEED] = self.fan_speed
+            data[ATTR_FAN_SPEED_LIST] = self.fan_speed_list
+
+        return data
+
+    def turn_on(self, **kwargs):
+        """Turn the vacuum on and start cleaning."""
+        raise NotImplementedError()
+
+    def async_turn_on(self, **kwargs):
+        """Turn the vacuum on and start cleaning.
+
+        This method must be run in the event loop and returns a coroutine.
+        """
+        return self.hass.async_add_job(partial(self.turn_on, **kwargs))
+
+    def turn_off(self, **kwargs):
+        """Turn the vacuum off stopping the cleaning and returning home."""
+        raise NotImplementedError()
+
+    def async_turn_off(self, **kwargs):
+        """Turn the vacuum off stopping the cleaning and returning home.
+
+        This method must be run in the event loop and returns a coroutine.
+        """
+        return self.hass.async_add_job(partial(self.turn_off, **kwargs))
+
+    def return_to_base(self, **kwargs):
+        """Set the vacuum cleaner to return to the dock."""
+        raise NotImplementedError()
+
+    def async_return_to_base(self, **kwargs):
+        """Set the vacuum cleaner to return to the dock.
+
+        This method must be run in the event loop and returns a coroutine.
+        """
+        return self.hass.async_add_job(partial(self.return_to_base, **kwargs))
+
+    def stop(self, **kwargs):
+        """Stop the vacuum cleaner."""
+        raise NotImplementedError()
+
+    def async_stop(self, **kwargs):
+        """Stop the vacuum cleaner.
+
+        This method must be run in the event loop and returns a coroutine.
+        """
+        return self.hass.async_add_job(partial(self.stop, **kwargs))
+
+    def locate(self, **kwargs):
+        """Locate the vacuum cleaner."""
+        raise NotImplementedError()
+
+    def async_locate(self, **kwargs):
+        """Locate the vacuum cleaner.
+
+        This method must be run in the event loop and returns a coroutine.
+        """
+        return self.hass.async_add_job(partial(self.locate, **kwargs))
+
+    def set_fan_speed(self, **kwargs):
```

Comment:
The methods in the base class should have the same signature as the overwritten methods in the child classes. It's good to break out the specific arguments as you've done in the xiaomi platform.
```py
def set_fan_speed(self, fan_speed, **kwargs):
```